### ISSUE TITLE
exactOptionsParameters - still forcibly inserting undefined with optiona into Tuple when using spread operator on sub Tuple

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#45764]
# Bug Report

### 🔎 Search Terms

- strictOptionalProperties  #43947 #44548
- exactOptionsParameters 
- [Labelled Tuple Elements](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#labeled-tuple-elements)
- Optional Tuple Parameters (enabled with labeling)
-  [Variadic Tuple Types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#variadic-tuple-types)

### 🕗 Version & Regression Information

- Spreading of Variadic Tuple types, as well as the ability to indicate final position optional parameters with labels,  were introduced in ts 4.0.  The noted behavior has been been present since then.

- However, although there was hope it would be fixed with 4.4's exact `exactOptionsParameters`, it was not.

### ⏯ Playground Link

https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true&ts=4.4.2#code/ATAuE8AcFNgFQIzALzANoA8BcA7ArgLYBG0ATgDTAYD8WRA9vQDbQCGOAugFAgDG9OAM6gwCBFkQp0CcnhwATaADMAljmjyOwAPTbgAUQAaAYX0AZM-oBycAIS3gAdxWgAFsGgZWvUAHlIoCoCrEwACqT0MKQQcFDQgligpHiwbiqCwCS8rASw6cDsHqQRpAB0wDzA-EIioAgATBJIqGgySSncfALCogDMTVKt3JUQMPD1gwDkAHqT5KULiOSTACSTnVXdtfXicBMtM3NtydDLaxwA3JXVPaD1jXtTs+Qycoqq6vJn6xc6ejj0MikYAAWmAxjMAEEAJIAWVBYFc+QAygAJXwAVTMABFMrBCkD6KRrlswPV+o8Ds8ZKsfn8isCwRCYfCwWkMmjMTjgFZfHA8QUcAyiRUgA

### 💻 Code

```ts
  type T1 = [x:number, x?:boolean]
  const t11:T1 = [1,undefined] // EXCELLENT!! with exactOptionalPropertyTypes:true this became is an error. 
  const t12:T1 = [1,true]
  const t13:T1 = [1]

  type T2 = ['^',...T1,'$']
  const t21:T2 = ['^',1,true,'$'];
  const t22:T2 = ['^',1,undefined,'$']; // noerr - CLAIM - this SHOULD be an error
  const t23:T2 = ['^',1,'$']; // err - CLAIM - this SHOULD NOT be an error 

```

### 🙁 Actual behavior

In the above code, variable declaration for `t22` is NOT an err.
In the above code, variable declaration for `t23` IS an err.

### 🙂 Expected behavior

In the above code, variable declaration for `t22` IS an err.
In the above code, variable declaration for `t23` is NOT an err.

Even though with `exactOptionalPropertyTypes` , the `T1` now no longer considers `[number,undefined]` to be a legal assignment,  when `T1` is spread into `T2` the old behavior is used instead of the new behavior.

The [documentation for exactOptionalPropertyTypes]() says 

> In TypeScript 4.4, the new flag --exactOptionalPropertyTypes specifies that optional property types should be interpreted exactly as written, meaning that `| undefined` is not added to the type:

Looking into the internal representation of T1 as a type, I see no change between 4.3.5 and 4.4.2 - `|undefined` is still added to the type in the case of 4.4.2.  That differs from what the documentation says - although the documentation doesn't really need to be concerned with internals - just the end result.

In both 4.3.5 and 4.4.2 the type internals looks something like this (sub-types pulled up  for clarity):
```
{
  "type": {
    "info": "this is a ts.Type with id=86"
  },
  "discTypeFlag": "Union",
  "c": {
    "ctypes": [
      {
        "discTypeFlag": "Undefined"
      },
      {
        "discTypeFlag": "BooleanLiteral",
        "value": false
      },
      {
        "discTypeFlag": "BooleanLiteral",
        "value": true
      }
    ]
  }
}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
{
  "type": {
    "info": "this is a ts.Type with id=85"
  },
  "objectSubType": "tuple",
  "discTypeFlag": "Object",
  "c": {
    "ctypes": [
      {
        "discTypeFlag": "Number"
      },
      {
        "$ref": {
          "info": "this is a ts.Type with id=86"
        },
        "discTypeFlag": "Union"
      }
    ]
  },
  "target": {
    "elementFlags": [
      [
        "Required"
      ],
      [
        "Optional"
      ]
    ],
    "hasRestElement": false,
    "fixedLength": 2,
    "minLength": 1,
    "combinedFlags": [
      "Required",
      "Optional"
    ],
    "readonly": false
  }
}
```

The type `T2` has 4 required places according to the internal rep shown here:
```
{
  "type": {
    "info": "this is a ts.Type with id=108"
  },
  "objectSubType": "tuple",
  "discTypeFlag": "Object",
  "c": {
    "ctypes": [
      {
        "discTypeFlag": "StringLiteral",
        "value": "^"
      },
      {
        "discTypeFlag": "Number"
      },
      {
        "$ref": {
          "info": "this is a ts.Type with id=86"
        },
        "discTypeFlag": "Union"
      },
      {
        "discTypeFlag": "StringLiteral",
        "value": "$"
      }
    ]
  },
  "target": {
    "elementFlags": [
      [
        "Required"
      ],
      [
        "Required"
      ],
      [
        "Required"
      ],
      [
        "Required"
      ]
    ],
    "hasRestElement": false,
    "fixedLength": 4,
    "minLength": 4,
    "combinedFlags": [
      "Required"
    ],
    "readonly": false
  }
}
```

**Comment by MartinJohns** [comment]
> However, although there was hope it would be fixed with 4.4's exact `exactOptionsParameters`, it was not.

This is the first time I ever heard of this flag. It's not mentioned in the 4.4 release notes. Where did you find this?

**Comment by craigphicks** [comment]
@MartinJohns - `exactOptionsPropertyNames`.

Despite the name, it apparently applies to Tuples as well.  From the link given in the 4.4 documentation [to the related pull](https://github.com/microsoft/TypeScript/pull/43947)

> In --strictOptionalProperties mode, optional elements in tuple types are strictly checked when they don't explicitly include 
> undefined in their type. For example:
> 
```
// Compile in --strictOptionalProperties mode
function f4(t: [string?]) {
    let x = t[0];  // string | undefined
    t[0] = 'hello';
    t[0] = undefined;  // Error, 'undefined' not assignable to 'string'
}
```
It appears the name changed from `strictOptionalProperties` to `exactOptionalPropertyTypes`.  (There never was a `strictOptionalParameters`.)

**Comment by craigphicks-public** [comment]
@andrewbranch - thank you.

**Comment by craigphicks** [comment]
@andrewbranch  thank you

**Comment by andrewbranch** [comment]
FWIW, that JSON dump of type info is not from our compiler API. I’ve never heard of properties like `discTypeFlag` and `ctypes`. It would be sort of helpful to me if this issue were focused on the tuple thing and the issue about what I assume is someone’s 3rd party compiler API were relocated somewhere else.

**Comment by craigphicks** [comment]
@andrewbranch - Sure - I'll take it out.  Should I remove that whole final section "Tangent topic yet critical - Type API may have changed - is it still possible to determine behavior from exposed Type API ?"  

It's not third party API or a compiler.  It's from code I wrote displaying the information obtained via `getTypeOfSymbolAtLocation` and `getDeclaredTypeOfSymbol`, which are part of the (bleeding edge) typescript compiler API.  That format you see is that information slightly massaged to take up less space, without losing anything critical. There is a verbose mode which outputs raw-ish information.

It's part of a generalized type explorer, which is a component of a typescript Type to schema translator, which is part of a program for implementing run time checking of Typescript types.  It all depends upon that (bleeding edge) Typescript compiler API (with entry points `getTypeOfSymbolAtLocation` and `getDeclaredTypeOfSymbol`) offering a consolidated Type which can be used to build a schema.

That final section of my post is about that (bleeding edge) Type API - perhaps - no longer enough info to predict what will pass the compiler in some instances (optional vs. explicit undefined).  That's important to me, because the built schemas are supposed to match the compiler (within some well defined bounds).

**Comment by craigphicks** [comment]
@andrewbranch - [Here is a link to the rawish output](https://gist.github.com/craigphicks/49b31c800a6ce2ef2a67006b84ec4d5a).  The `discTypeFlag` is still there but it is just taken from `ts.Type['flags']`.  Sometimes (rarely) more than one flag bit is set,  but an unambiguous value is required for discriminating types for processing. 

In the out you see some places where `r:` and c:` fields exist side by side.  `r` is for raw`.  `c` is the massaged and shortened form, which may be ignored.

**Comment by andrewbranch** [comment]
Yeah, it would be better if an API request were split into a separate issue please. And it would be easier for me to evaluate without it being intercepted by “a generalized type explorer, which is a component of a typescript Type to schema translator.” If you’re not using the compiler API directly, I can’t tell whether the compiler API is actually deficient.

**Comment by craigphicks** [comment]
@andrewbranch   Removed.  

`getTypeOfSymbolAtLocation` and `getDeclaredTypeOfSymbol` are both member functions of checker, and they are both exposed as part of the compiler API.  For example, `getTypeOfSymbol` is [introduced on the Typescript github page "Using the Compiler API"](https://github.com/microsoft/TypeScript-wiki/blob/main/Using-the-Compiler-API.md).  

But I think that is not the Compiler API you mean right now.  You mean using the compiler API exposed either through `tsc` or `ts.createProgram`, `ts.getXXXDiagnostics` and `program.emit`, which is 99.9% of usage. I agree 100%.

**Comment by andrewbranch** [comment]
No, I’m familiar with the full checker API, and I’m sympathetic to ensuring it’s usable, it just doesn’t seem relevant to this issue and I want to evaluate it without it having been serialized into a different structure with different property names by some mysterious 3rd party tool.

**Comment by craigphicks** [comment]
@andrewbranch - You are absolutely correct that it is a separate issue and my judgement was poor. (I've already figure how to solve that separate API issue so don't worry).

This cause of (and solution) of this current issue is illustrated in the following example:

```
  type T3 = [x:number, x:boolean]|[x:number]
  type T4 = ['^',...T3,'$']
  const t41:T4 = ['^',1,true,'$'];
  const t42:T4 = ['^',1,undefined,'$']; // fails - should fail and it does
  const t43:T4 = ['^',1,'$']; // should pass and it does 
```

Why this works and `T1`/`T2` above doesn't work is that when the compiler sees the union it pulls the union up the top level, so we actually get a union of 
```
type T4eq = ['^',...[x:number, x:boolean],'$'] | ['^',...[x:number],'$']  
```
So some extra code to convert a tuple with (an) optional parameter(s) into a union of N+1 types, where N is the number of optional parameters, would fix it, I guess.

**Comment by Andarist** [comment]
The OP claims that the type produced by this spread is incorrect:
```ts
type T1 = [x: number, x?: boolean];
type T2 = ["^", ...T1, "$"]; // ["^", number, boolean | undefined, "$"]
```
The latter examples imply that the expected behavior would be to produce `["^", number, boolean?, "$"]` but that is an impossible type (`A required element cannot follow an optional element.(1257)`). 

When spreading a tuple without adding extra trailing elements the original optional flag is preserved (and not expanded into `T | undefined`). So I'm really not sure if there is anything actionable here. I think that the issue could be closed.

**Comment by RyanCavanaugh** [comment]
> ... but that is an impossible type

Yeah, there are a couple things wrapped up into one here

 * Should tuples support middle optional elements?
 * If not, what should a spread do if doing one would produce one?

The first one is effectively a feature request, but not one we'd be likely to pick up since I think it'd have serious combinatorial explosion problems.

Re second one, the current behavior is maybe the best approximation of what we could do. I think there's an argument to be made that under EOPT, tuple spreads should behave as if you had written
```ts
  type T2 = ['^',...Required<T1>,'$'];
  const t21:T2 = ['^',1,true,'$'];
  const t22:T2 = ['^',1,undefined,'$']; // should be + is an error
```
Though if you want that behavior, you *can* write that today, but if you like today's behavior more and we switch, there wouldn't be a way to go back except to turn off EOPT -- but maybe that's what EOPT *means*, after all.

**Comment by RyanCavanaugh** [comment]
Having just read what I wrote, I'm maybe unsure now, since it's not like `T2` acquires a *wrong* type here. It's maybe surprising, but the definition doesn't introduce a novel soundness hole or anything.
